kclvm/compiler_base/3rdparty/rustc_errors/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+//! Text rendering and related helper functions.
+//!
+//! Reuse 'styled_buffer.rs' in 'rustc_errors',
+//! and 'styled_buffer.rs' has been modified to fit the feature of 'Compiler-Base'.
+//!
+//! - add method 'appendl()' and 'putl()' to 'StyledBuffer'.
+//! - replaced the 'enum Style' with 'trait Style' to support extending more styles.
+//! - add some test cases for 'StyledBuffer'.
+
+use std::any::Any;
+use termcolor::ColorSpec;
+
+pub mod styled_buffer;
+
+/// 'Style' is a trait used to specify the 'enum Style' supported by 'StyleBuffer'.
+///
+/// It provides the following 4 methods.
+///
+/// as_any(&self) : return 'self' to support for downcasting concrete style types.
+/// box_clone(&self) : copy 'self' and return via 'Box' wrapper.
+/// style_eq(&self, other: &Box<dyn Style>) : downcast 'other' to concrete style type and compare equivalence with 'self'.
+/// fn render_style(&self) : render style to terminal color/font configuration.
+///
+/// # Examples
+///
+/// ```rust
+/// use rustc_errors::Style;
+/// use termcolor::{ColorSpec, Color};
+/// use core::any::Any;
+///
+/// #[derive(Copy, Clone, Debug, PartialEq, Hash)]
+/// pub enum DummyStyle{
+///     Dummy
+/// }
+///
+/// impl Style for DummyStyle{
+///    // return 'self' to support for downcasting concrete style types.
+///    fn as_any(&self) -> &dyn Any{
+///        self
+///    }
+///
+///    // copy 'self' and return via 'Box' wrapper.
+///    fn box_clone(&self) -> Box<dyn Style> {
+///         Box::new((*self).clone())
+///    }
+///
+///    fn render_style(&self) -> ColorSpec {
+///         let mut spec = ColorSpec::new();
+///         match self{
+///             DummyStyle::Dummy => {
+///                 spec.set_fg(Some(Color::Red)).set_intense(true);
+///             }
+///         }
+///         spec
+///    }
+///
+///    fn style_eq(&self, other: &Box<dyn Style>) -> bool {
+///         let other_style: &DummyStyle = match other.as_any().downcast_ref::<DummyStyle>() {
+///             Some(sty) => sty,
+///             None => panic!("Err")
+///         };
+///         *self == *other_style
+///    }
+/// }
+/// ```
+pub trait Style {
+    fn as_any(&self) -> &dyn Any;
+    fn box_clone(&self) -> Box<dyn Style>;
```

Comment:
`box_clone()` may have a default impl.  A further question is do users need to override the box_clone() method? If not, can it be taken out of this method as a constraint on the `Style` trait.

Reply:
replaced the `enum Style` with generics `T: Clone + PartialEq + Eq + Style` to support extending more styles.